integration/benchmark_test.go

```diff
@@ -0,0 +1,239 @@
+			expectedFiles: []string{"var/lib/rpm/Packages", "etc/lsb-release", "etc/os-release", "/config", "usr/lib/os-release"},
+			expectedOS:    analyzer.OS{Name: "1.0", Family: "photon"},
+		},
+		{
+			name:          "happy path, registry.redhat.io/ubi7",
+			imageName:     "registry.redhat.io/ubi7",
+			imageFile:     "testdata/fixtures/ubi-7.tar.gz",
+			expectedFiles: []string{"etc/redhat-release", "etc/system-release", "/config", "var/lib/rpm/Packages", "etc/os-release"},
+			expectedOS:    analyzer.OS{Name: "7.7", Family: "redhat"},
+		},
+		{
+			name:          "happy path, opensuse leap 15.1",
+			imageName:     "opensuse/leap:latest",
+			imageFile:     "testdata/fixtures/opensuse-leap-151.tar.gz",
+			expectedFiles: []string{"usr/lib/os-release", "usr/lib/sysimage/rpm/Packages", "/config", "etc/os-release"},
+			expectedOS:    analyzer.OS{Name: "15.1", Family: "opensuse.leap"},
+		},
+		{
+			name:                 "happy path, vulnimage with lock files",
+			imageName:            "knqyf263/vuln-image:1.2.3",
+			imageFile:            "testdata/fixtures/vulnimage.tar.gz",
+			expectedFiles:        []string{"etc/os-release", "node-app/package-lock.json", "python-app/Pipfile.lock", "ruby-app/Gemfile.lock", "rust-app/Cargo.lock", "/config", "etc/alpine-release", "lib/apk/db/installed", "php-app/composer.lock"},
+			expectedOS:           analyzer.OS{Name: "3.7.1", Family: "alpine"},
+			expectedLibraries:    "testdata/goldens/knqyf263vuln-image:1.2.3.expectedlibs.golden",
+			expectedPkgsFromCmds: "testdata/goldens/knqyf263vuln-image:1.2.3.expectedpkgsfromcmds.golden",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc // save a copy of tc for use in t.Run https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721
+		b.Run(tc.name, func(b *testing.B) {
+			//b.Parallel()
+			ctx := context.Background()
+			d, _ := ioutil.TempDir("", "TestFanal_Library_*")
+			defer os.RemoveAll(d)
+			c := cache.Initialize(d)
+
+			opt := types.DockerOption{
+				Timeout:  600 * time.Second,
+				SkipPing: true,
+			}
+
+			cli, err := client.NewClientWithOpts(client.FromEnv)
+			require.NoError(b, err, tc.name)
+
+			// ensure image doesnt already exists
+			_, _ = cli.ImageRemove(ctx, tc.imageFile, dtypes.ImageRemoveOptions{
+				Force:         true,
+				PruneChildren: true,
+			})
+
+			testfile, err := os.Open(tc.imageFile)
+			require.NoError(b, err)
+
+			// load image into docker engine
+			_, err = cli.ImageLoad(ctx, testfile, true)
+			require.NoError(b, err, tc.name)
+
+			// tag our image to something unique
+			err = cli.ImageTag(ctx, tc.imageName, tc.imageFile)
+			require.NoError(b, err, tc.name)
+
+			ext, err := docker.NewDockerExtractor(opt, c)
+			require.NoError(b, err, tc.name)
+			ac := analyzer.Config{Extractor: ext}
+
+			b.ReportAllocs()
+			b.ResetTimer()
+			// run tests twice, one without cache and with cache
+			for i := 1; i <= 2; i++ {
+				runChecksBench(5, b, ac, ctx, tc, d, c)
+
+			}
```

Comment:
This sums two results, with and without cache, right? This is useful for integration tests, but not useful for benchmarks. We need to know those scores respectively. 

Commit A
- With cache: 1000ms
- Without cache: 10ms
- Total: 1010ms

Commit B
- With cache: 1000ms
- Without cache: 50ms
- Total: 1050ms

The total looks almost the same (1000ms => 1050ms), but the score without cache gets much worse (10ms => 50ms). If we see only the total, we must think it is fine.

Reply:
Yes I had asked similar questions here https://github.com/aquasecurity/fanal/issues/69 I wasn't sure what direction we wanted to head in so I've left it open for comments.

Reply:
There was also a bug in this benchmark code that I found, I think these results are much more meaningful:

```
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_alpine:3.10-8                      19          58401294 ns/op        11952605 B/op     156570 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_amazonlinux:2-8                     1        3391656811 ns/op        36893008 B/op     566862 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_debian:buster-8                     1        3209189846 ns/op         7467096 B/op     228334 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_photon:1.0-8                        1        3694983184 ns/op        16589568 B/op     397234 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_registry.redhat.io/ubi7-8                   1        6131400243 ns/op        65793568 B/op     831039 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_opensuse_leap_15.1-8                        1        2519825637 ns/op        23268648 B/op     453900 allocs/op
BenchmarkFanal_Library_DockerMode_WithoutCache/happy_path,_vulnimage_with_lock_files-8                 1        4892505695 ns/op        24762120 B/op     414372 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_alpine:3.10-8                                 18          59800578 ns/op        11953707 B/op     156572 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_amazonlinux:2-8                                8         132578903 ns/op        36849505 B/op     566711 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_debian:buster-8                               13          83306147 ns/op         7420395 B/op     228188 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_photon:1.0-8                                  10         106737022 ns/op        16565464 B/op     397098 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_registry.redhat.io/ubi7-8                      6         170368824 ns/op        65745706 B/op     830901 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_opensuse_leap_15.1-8                          16          71487653 ns/op        23241808 B/op     453769 allocs/op
BenchmarkFanal_Library_DockerMode_WithCache/happy_path,_vulnimage_with_lock_files-8                    7         155990412 ns/op        24704204 B/op     413966 allocs/op

```

Reply:
Oh, I'm sorry not to notice #69.